### PR TITLE
Add workflow API methods to App API, versioning to Project API

### PR DESCRIPTION
**Pull request by GoldenAnpu**
### Add workflow API methods to App API

The workflow functionality is used to create connections between the states of projects and tasks (application sessions) that interact with them in some way. By assigning connections to various entities, the workflow tab allows tracking the history of project changes. This functionality uses versioning (optional).
The active task always acts as a node, for which input and output elements are defined. There can be multiple input and output elements. A task can also be used as an input or output element. For example, an inference task takes a deployed model and a project as inputs, and the output is a new state of the project.

1. **app_api.py**
   - add workflow methods

2. **module_api.py**
   - add `ApiField.NODE`

3. **fs.py**
   - add def `str_is_url` 

4. **io/json.py**
    - add def `validate_json` to validate data with predefined schema

5. **output/__init__.py**
    - refactor def `set_download` to return FileInfo object and replace `print` with `logger.info`

----

### Add versioning to Project API

Versioning of projects is used for creating backup versions of projects and subsequently storing them in the `system` folder within team Files. At the level of each project, there is a mapping file named `versions.json` to ensure the autonomy of versions. Each version archive includes a timestamp of its creation in its name and is mapped in `versions.json`, as well as in the version information that can be obtained for the project via an API request. Inside each archive, there is a `.tar` file compressed with `zstd` containing a `bin` file created using `pickle` (what exactly and in what order can be learned from a docstring). This file stores information about all the Info that can be obtained for the project.

These data are unpacked in a convenient format, and the project is restored through mapping old elements to newly created ones with the same parameters. When restoring a version, a new project is created.

1. **entity_annotation/tag_api.py**
    - refactor def `append_to_objects` to return ids and update its docstring
    - add def `add_to_objects` with restriction (works only with Image projects)

2. **module_api.py**
   - add `ApiField.COMMIT_TOKEN`, `ApiField.TEAM_FILE_ID`, `ApiField.PROJECT_UPDATED_AT`, `ApiField.OPERATOR`
  
3.  **project_api.py**
    - add `DataVersion` as `self.version` to `__init__` for `ProjectAPI` 
    - add `version` field to `ProjectInfo`
    - exclude `version` field for listing requests

4. **project/__init__.py**
    - add DataVersion import
    - auto formatting of code lines for error string

5. **project/data_version.py** - new
    - add classes `VersionInfo` and `DataVersion` with its method to cover versioning functionality

6. **project/project.py**
    - add static methods `download_bin` and `upload_bin` to cover versioning functionality needs, such as the new project storage format
    - update rtype for `find_project_dirs`
    - fix pylint errors for `_download_dataset`

----

### Other updates
1. **advanced_api.py**
    - fix `Optional[str or int]` in typing
    
2. **api.py**
    - just update instance URL in docstrings

3. **setup.py**
    - add `zstd` module for `.tar` files compression

7. **figure_api.py**
    - add def `create_bulk`

8. **file_api.py**
    - update def `exists` and `dir_exists` with `recursive` arg to speed up checks
    - add def  `get_json_file_content`

9. **image_api.py**
    - add to def `remove_batch` more info in docstring
    - add def `check_existing_links` to check remote images like hashes

10. **task_api.py**
    - add def `set_output_text` to create custom output messages
    - update docstring in `get_info_by_id` to show example of app info in meta

### CODE REVIEW

File: supervisely/api/entity_annotation/figure_api.py [hunk 1]
```diff
@@ -373,6 +373,45 @@ def _append_bulk(
                 figure_id = resp_obj[ApiField.ID]
                 key_id_map.add_figure(key, figure_id)
 
+    def create_bulk(
+        self,
+        figures_json: List[dict],
+        entity_id: int = None,
+        dataset_id: int = None,
+        batch_size=200,
+    ) -> List[int]:
```
Comment (NikolaiPetukhov) [inline supervisely/api/entity_annotation/figure_api.py:382]: Maybe also add to api.annotation as add_labels_bulk or smth

File: supervisely/api/image_api.py [hunk 2]
```diff
@@ -920,6 +922,39 @@ def check_existing_hashes(
...
+        results = []
+        if len(links) == 0:
+            return results
+        for links_batch in batched(links, batch_size=900):
+            with ThreadPoolExecutor(max_workers=20) as executor:
+                b_results = list(executor.map(_is_image_available, links))
+            results.extend(b_results)
+
+            if progress_cb is not None:
+                progress_cb(len(links_batch))
```
Comment (NikolaiPetukhov) [inline supervisely/api/image_api.py:955]: Can remove batched and move progress_cb call into _is_image_available

**Review by NikolaiPetukhov — APPROVED** [review]

File: supervisely/project/data_version.py [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+        :param do_initialization: Initialize project versions. Set to False for internal use.
+        :type do_initialization: bool
+        :return: Project versions
+        :rtype: dict
+        """
+        if do_initialization:
+            self.initialize(project_info)
+        file_info = self._api.file.get_info_by_path(self.project_info.team_id, self.versions_path)
+        if file_info:
+            response = requests.get(file_info.full_storage_url, stream=True)
```
[the remark this answers is not in the document]
Comment (GoldenAnpu) [reply supervisely/project/data_version.py:150]: Refactor method to get JSON content

File: supervisely/api/entity_annotation/figure_api.py (hunk 1 above)
Comment (GoldenAnpu) [reply supervisely/api/entity_annotation/figure_api.py:382]: Will be added in another request

File: supervisely/project/project.py [hunk 4]
```diff
@@ -2673,24 +2673,23 @@ def download_bin(
...
-                    progress_cb(len(batch))
-        dataset_infos = [ds for ds in dataset_infos if ds.id not in dataset_skip_list]
-        if return_bytesio:
-            data = io.BytesIO()
-            pickle.dump(
-                (project_info, meta, dataset_infos, image_infos, figures, alpha_geometries), data
-            )
-            return data
+                if log_progress or progress_cb is not None:
+                    ds_progress(len(batch))
```
Comment (NikolaiPetukhov) [inline supervisely/project/project.py:2677]: if both log_progress and progress_cb are present, then progress_cb will never be called

File: supervisely/api/entity_annotation/figure_api.py [hunk 5]
```diff
@@ -373,6 +373,45 @@ def _append_bulk(
...
+            body = {ApiField.DATASET_ID: dataset_id}
+        else:
+            body = {ApiField.ENTITY_ID: entity_id}
+
+        for batch_jsons in batched(figures_json, batch_size):
+            body[ApiField.FIGURES] = batch_jsons
+            response = self._api.post("figures.bulk.add", body)
+            for resp_obj in response.json():
+                figure_ids.append(resp_obj[ApiField.ID])
+        return figure_ids
```
Comment (NikolaiPetukhov) [inline supervisely/api/entity_annotation/figure_api.py:413]: Chek if geometry is alpha mask and upload it separately

File: supervisely/project/project.py (hunk 4 above)
Comment (GoldenAnpu) [reply supervisely/project/project.py:2677]: Add a higher priority for progress_cb

**Pull request merged**